common/hashing.h

```diff
@@ -474,6 +481,63 @@ inline auto CarbonHashValue(const llvm::SmallString<Length>& value,
   return CarbonHashValue(llvm::StringRef(value.data(), value.size()), seed);
 }
 
+// Support types that are array like by building an `llvm::ArrayRef` out of
+// them. We can't do this by accepting any type convertible to an `ArrayRef`
+// because that type supports building a synthetic array out of any single
+// element. These need to go back through the dispatch implementation though to
+// pick up the full space of hash overloads for `T`.
+template <typename T>
+inline auto CarbonHashValue(const std::vector<T>& arg, uint64_t seed)
+    -> HashCode {
+  return DispatchImpl(llvm::ArrayRef(arg), seed);
+}
+template <typename T>
+inline auto CarbonHashValue(const llvm::SmallVectorImpl<T>& arg, uint64_t seed)
+    -> HashCode {
+  return DispatchImpl(llvm::ArrayRef(arg), seed);
+}
+template <typename T, size_t N>
+inline auto CarbonHashValue(const std::array<T, N>& arg, uint64_t seed)
+    -> HashCode {
+  return DispatchImpl(llvm::ArrayRef(arg), seed);
+}
+template <typename T, size_t N>
+inline auto CarbonHashValue(const T (&arg)[N], uint64_t seed) -> HashCode {
+  return DispatchImpl(llvm::ArrayRef(arg), seed);
+}
+
+inline auto CarbonHashValue(llvm::APInt value, uint64_t seed) -> HashCode {
+  Hasher hasher(seed);
+  if (LLVM_LIKELY(value.isSingleWord())) {
+    hasher.Hash(value.getBitWidth(), value.getZExtValue());
```

Comment:
I suspect the answer is no, but is it worth treating the `getBitWidth() <= 32` case as a special case, for which we only need to hash 8 bytes rather than 12?